### ISSUE TITLE
a11y: Improve keyboard focus navigation in main window

### ISSUE THREAD

**Issue by jcsteh** [mumble-voip/mumble#2291]
In Mumble's main window, once keyboard focus hits the chat input text box, it's not possible to move focus to other areas of the main window such as the channels tree or log text. Tab and shift+tab are both intercepted by the text box. I understand why tab is intercepted. There are a couple of ways to solve this:

1. Allow shift+tab to move focus to the previous control. While this means tabbing is asymmetric, it's a common technique where there is a large text box that needs to intercept the tab key. For example, when composing a message in Mozilla Thunderbird, once you tab into the message body, you can't "tab" out of it, but you *can* "shift+tab" out of it.
2. Provide another key to move focus to other areas in the main window. F6 is often used to switch between panes or panels. For example, when in the chat input box, f6 might take you to the channels tree.

A related issue is that if there are links, etc. in the log, there can be several tab stops between the log and the chat input box. This makes for an inefficient user experience for keyboard users. This could be solved in one of two ways:

1. Like 2) above, provide a key such as f6 to move between main areas. So, f6 might cycle between the channels tree and the chat input field.
2. Provide shortcut keys to move to specific controls; e.g. alt+shift+i for the chat input box, alt+shift+c for the channels tree and alt+shift+l for the log. This make sit very easy to get to specific controls, but is less discoverable for new users.

These issues are particularly problematic for blind screen reader users, as such users usually rely entirely on the keyboard.

**Comment by mkrautz** [comment]
All of the potential solutions sound sane to me, but I don't know enough about focus handling in Qt to make any educated comments on it.

We just need to figure out the correct layer for this. Is this something Mumble should handle, or is it possible for Mumble to provide focus hints to Qt, to make it handle some of it? Getting this stuff right by listening to key events and manually switching the focused element seems fragile to me. (But, as I said, I don't know enough about Qt's focus handling yet...).

@jcsteh Thanks for filing these issues. They really help. Especially the suggestions on how to solve the problems. Thank you.

**Comment by jcsteh** [comment]
I don't have any experience with QT focus handling either. I'm not sure if you've read this already, but I found [this QT documentation about keyboard focus](http://doc.qt.io/qt-5/focus.html).

* It suggests that control+tab should work as tab in editors, but it doesn't seem to here. Eve if it did, control+tab normally switches tab sheets, so that's a very strange approach.
* It does suggest editors should override `QWidget::event` for special tab handling. You could do this for the shift+tab approach.
* There doesn't seem to be any specific support for f6.
* You can assign keyboard accelerators to labels. However, if I'm not mistaken, there are no labels for the controls in the main window, so this isn't really an option.

So, I think custom key event handling is unfortunately the only way this can be done.

**Comment by mkrautz** [comment]
- [ ] Implement Shift-Tab to move focus backwards, when inside the text widget in MainWindow
- [ ] Implement F6 to switch between major areas in MainWindow (log view, text view, server view)

**Comment by mkrautz** [comment]
Shift-Tab for the chat input bar implemented in https://github.com/mumble-voip/mumble/pull/2410
F6 for MainWindow implemented in https://github.com/mumble-voip/mumble/pull/2411

Both pending review...

**Comment by mkrautz** [comment]
Both of these are available in the latest snapshot.
